File: lib/features/agents/services/safety_agent.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:cloud_functions/cloud_functions.dart';

/// Safety Agent provider
final safetyAgentProvider = Provider<SafetyAgent>((ref) {
  return SafetyAgent();
});

/// Safety Agent - Enhanced Content Safety
/// 
/// Features:
/// - Content moderation
/// - Behavior analysis
class SafetyAgent {
  final FirebaseFunctions _functions = FirebaseFunctions.instance;

  /// Moderate content
  Future<ModerationResult> moderateContent({
    required ContentType contentType,
    required String content,
    required String userId,
    required String location,
    int previousViolations = 0,
    String language = 'en',
  }) async {
    try {
      final callable = _functions.httpsCallable('moderateContent');
      final result = await callable.call<Map<String, dynamic>>({
        'contentType': contentType.name,
        'content': content,
        'context': {
          'userId': userId,
          'location': location,
          'previousViolations': previousViolations,
        },
        'language': language,
      });
      return ModerationResult.fromJson(result.data);
    } catch (e) {
      // Fail safe - allow if moderation fails
      return ModerationResult.safe();
    }
  }

  /// Analyze user behavior for safety concerns
  Future<BehaviorAnalysis> analyzeBehavior({
    required String userId,
    required List<UserAction> recentActions,
    List<ReportRecord>? reportHistory,
  }) async {
    try {
      final callable = _functions.httpsCallable('analyzeBehavior');
      final result = await callable.call<Map<String, dynamic>>({
        'userId': userId,
        'recentActions': recentActions.map((a) => a.toJson()).toList(),
        if (reportHistory != null)
          'reportHistory': reportHistory.map((r) => r.toJson()).toList(),
      });
      return BehaviorAnalysis.fromJson(result.data);
    } catch (e) {
      return BehaviorAnalysis.normal();
    }
  }

  /// Quick check if text is safe
  Future<bool> isTextSafe(String text, {String location = 'chat'}) async {
    final result = await moderateContent(
      contentType: ContentType.text,
      content: text,
      userId: 'quick_check',
      location: location,
    );
    return result.isSafe;
  }
}

// ==================== ENUMS ====================

enum ContentType { text, username, bio, story, comment }
enum SeverityLevel { none, low, medium, high, critical }
enum ModerAction { allow, warn, filter, block, escalate }
enum RiskLevel { low, medium, high, critical }
enum MonitoringLevel { normal, increased, strict }

// ==================== DATA MODELS ====================

class ModerationResult {
  final bool isSafe;
  final List<ModerationCategory> categories;
  final ModerAction suggestedAction;
  final String? filteredContent;
  final String explanation;

  ModerationResult({
    required this.isSafe,
    required this.categories,
    required this.suggestedAction,
    this.filteredContent,
    required this.explanation,
  });

  factory ModerationResult.fromJson(Map<String, dynamic> json) {
    return ModerationResult(
      isSafe: json['isSafe'] as bool? ?? true,
      categories: (json['categories'] as List<dynamic>?)
          ?.map((c) => ModerationCategory.fromJson(c))
          .toList() ?? [],
      suggestedAction: ModerAction.values.firstWhere(
        (a) => a.name == json['suggestedAction'],
        orElse: () => ModerAction.allow,
      ),
      filteredContent: json['filteredContent'] as String?,
      explanation: json['explanation'] as String? ?? '',
    );
  }

  factory ModerationResult.safe() {
    return ModerationResult(
      isSafe: true,
      categories: [],
      suggestedAction: ModerAction.allow,
      explanation: 'Content is safe',
    );
  }
}

class ModerationCategory {
  final String category;
  final SeverityLevel severity;
  final double confidence;

  ModerationCategory({
    required this.category,
    required this.severity,
    required this.confidence,
  });

  factory ModerationCategory.fromJson(Map<String, dynamic> json) {
    return ModerationCategory(
      category: json['category'] as String? ?? '',
      severity: SeverityLevel.values.firstWhere(
        (s) => s.name == json['severity'],
        orElse: () => SeverityLevel.none,
      ),
      confidence: (json['confidence'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class UserAction {
  final String action;
  final String timestamp;
  final String? target;

  UserAction({required this.action, required this.timestamp, this.target});

  Map<String, dynamic> toJson() => {
    'action': action,
    'timestamp': timestamp,
    if (target != null) 'target': target,
  };
}

class ReportRecord {
  final String reason;
  final String outcome;
  final String date;

  ReportRecord({required this.reason, required this.outcome, required this.date});

  Map<String, dynamic> toJson() => {
    'reason': reason,
    'outcome': outcome,
    'date': date,
  };
}

class BehaviorAnalysis {
  final RiskLevel riskLevel;
  final List<String> concerns;
  final String recommendedAction;
  final MonitoringLevel monitoringLevel;

  BehaviorAnalysis({
    required this.riskLevel,
    required this.concerns,
    required this.recommendedAction,
    required this.monitoringLevel,
  });

  factory BehaviorAnalysis.fromJson(Map<String, dynamic> json) {
    return BehaviorAnalysis(
      riskLevel: RiskLevel.values.firstWhere(
        (r) => r.name == json['riskLevel'],
        orElse: () => RiskLevel.low,
      ),
      concerns: List<String>.from(json['concerns'] ?? []),
      recommendedAction: json['recommendedAction'] as String? ?? 'No action needed',
      monitoringLevel: MonitoringLevel.values.firstWhere(
        (m) => m.name == json['monitoringLevel'],
        orElse: () => MonitoringLevel.normal,
      ),
    );
  }

  factory BehaviorAnalysis.normal() {
    return BehaviorAnalysis(
      riskLevel: RiskLevel.low,
      concerns: [],
      recommendedAction: 'No action needed',
      monitoringLevel: MonitoringLevel.normal,
    );
  }
}

File: functions/src/agents/safety/safety_agent.ts
/**
 * Safety Agent - Enhanced Content Safety
 * 
 * Features:
 * - Advanced content moderation with Tree of Thoughts reasoning
 * - Real-time toxicity detection with context awareness
 * - User behavior analysis with pattern detection
 * - Report handling with stepwise evaluation
 * 
 * Enhanced with Tree of Thoughts (ToT) for deliberate, multi-step moderation
 */

import { genkit, z } from 'genkit';
import { googleAI, gemini20FlashExp } from '@genkit-ai/googleai';
import { onCall, HttpsError } from 'firebase-functions/v2/https';
import { stepwiseToT } from '../tot/tot_utils';

const ai = genkit({
    plugins: [googleAI()],
    model: gemini20FlashExp,
});

// =============================================================================
// SCHEMAS
// =============================================================================

const ContentModerationInputSchema = z.object({
    contentType: z.enum(['text', 'username', 'bio', 'story', 'comment']),
    content: z.string(),
    context: z.object({
        userId: z.string(),
        location: z.string(),
        previousViolations: z.number().default(0),
    }),
    language: z.string().default('en'),
});

const ContentModerationOutputSchema = z.object({
    isSafe: z.boolean(),
    categories: z.array(z.object({
        category: z.string(),
        severity: z.enum(['none', 'low', 'medium', 'high', 'critical']),
        confidence: z.number(),
    })),
    suggestedAction: z.enum(['allow', 'warn', 'filter', 'block', 'escalate']),
    filteredContent: z.string().optional(),
    explanation: z.string(),
});

const BehaviorAnalysisInputSchema = z.object({
    userId: z.string(),
    recentActions: z.array(z.object({
        action: z.string(),
        timestamp: z.string(),
        target: z.string().optional(),
    })),
    reportHistory: z.array(z.object({
        reason: z.string(),
        outcome: z.string(),
        date: z.string(),
    })).optional(),
});

const BehaviorAnalysisOutputSchema = z.object({
    riskLevel: z.enum(['low', 'medium', 'high', 'critical']),
    concerns: z.array(z.string()),
    recommendedAction: z.string(),
    monitoringLevel: z.enum(['normal', 'increased', 'strict']),
});

// =============================================================================
// GENKIT FLOWS
// =============================================================================

/**
 * Content Moderation - Enhanced with Tree of Thoughts
 * 
 * Stepwise analysis with backtracking:
 * Step 1: Explicit violation check (slurs, threats, explicit content)
 * Step 2: Context-aware analysis (gaming context makes some terms safe)
 * Step 3: Severity assessment with backtracking if context changes interpretation
 */
export const moderateContentFlow = ai.defineFlow(
    {
        name: 'moderateContent',
        inputSchema: ContentModerationInputSchema,
        outputSchema: ContentModerationOutputSchema,
    },
    async (input) => {
        // Define moderation steps for stepwise ToT
        const steps = [
            {
                name: 'Explicit Violation Check',
                prompt: `STEP 1: Check for EXPLICIT violations in this ${input.contentType}.

Scan for:
- Slurs or hate speech (racial, gender, religious)
- Explicit threats of violence
- Adult/sexual content
- Spam patterns (URLs, repeated text)
- Personal information exposure (phone, address, SSN)
- Self-harm content

This is a FIRST PASS - be thorough but note that context will be evaluated next.
If no explicit violations found, mark as passed.
If violations found, note severity (low/medium/high/critical).`,
            },
            {
                name: 'Context-Aware Analysis',
                prompt: `STEP 2: Analyze content IN CONTEXT.

This content appears in: ${input.context.location}
Content type: ${input.contentType}
Language: ${input.language}

IMPORTANT - Gaming Context Rules:
- "kill", "destroy", "attack", "die" in gaming context = SAFE (e.g., "I'll kill you in the next round")
- Competitive trash talk between players = Usually SAFE if not personal attacks
- "noob", "gg", "rekt" = Gaming slang, SAFE

IMPORTANT - Backtracking:
- If previous step flagged something that is actually gaming context, mark shouldBacktrack=true
- Example: "I'm going to destroy you" flagged as threat → In game chat → Actually safe

Consider user's violation history: ${input.context.previousViolations} previous violations
${input.context.previousViolations > 2 ? 'WARNING: Repeat offender - be stricter' : ''}`,
            },
            {
                name: 'Severity Assessment',
                prompt: `STEP 3: Final severity assessment.

Based on Steps 1 and 2, determine:
1. Is content ultimately SAFE or UNSAFE?
2. What is the final severity level?
3. What action should be taken?

Action guidelines:
- allow: Clearly safe content
- warn: Borderline content, educate user
- filter: Partially unsafe, can be cleaned (e.g., replace bad words with ***)
- block: Clearly unsafe, do not show
- escalate: Ambiguous, needs human review

If context analysis (Step 2) cleared a flagged item, prefer 'allow' or 'warn'.
Consider: Is blocking this content worth potentially upsetting a user who was just using gaming terms?`,
            },
        ];

        const contextInfo = {
            contentType: input.contentType,
            location: input.context.location,
            userId: input.context.userId,
            previousViolations: input.context.previousViolations,
            language: input.language,
            isGamingContext: ['game', 'match', 'table', 'room'].some(g =>
                input.context.location.toLowerCase().includes(g)
            ),
        };

        try {
            // Run stepwise ToT moderation
            const { result, stepResults, shouldBlock } = await stepwiseToT(
                input.content,
                steps,
                contextInfo,
                ContentModerationOutputSchema
            );

            // Override if stepwise analysis determined block
            if (shouldBlock && result.suggestedAction === 'allow') {
                return {
                    ...result,
                    isSafe: false,
                    suggestedAction: 'block' as const,
                };
            }

            return result;
        } catch (error) {
            // Fallback to simple moderation if ToT fails
            console.error('ToT moderation failed, using fallback:', error);

            const prompt = `You are the Safety Agent for ClubRoyale.

Moderate this ${input.contentType}:
"${input.content}"

Context:
- Location: ${input.context.location}
- Previous violations: ${input.context.previousViolations}
- Language: ${input.language}

Check for:
1. Hate speech or harassment
2. Adult/explicit content
3. Violence references
4. Spam or scams
5. Personal information
6. Impersonation
7. Self-harm
8. Illegal content

Be strict but fair. Consider:
- Gaming context (competitive trash talk may be acceptable)
- Cultural differences
- Intent vs. impact

Provide action: allow (safe), warn (borderline), filter (partially unsafe), block (unsafe), escalate (needs human review).`;

            const { output } = await ai.generate({
                prompt,
                output: { schema: ContentModerationOutputSchema },
            });

            return output ?? {
                isSafe: true,
                categories: [],
                suggestedAction: 'allow' as const,
                explanation: 'Default safe',
            };
        }
    }
);

export const analyzeBehaviorFlow = ai.defineFlow(
    {
        name: 'analyzeBehavior',
        inputSchema: BehaviorAnalysisInputSchema,
        outputSchema: BehaviorAnalysisOutputSchema,
    },
    async (input) => {
        const prompt = `Analyze user behavior for safety concerns.

User: ${input.userId}
Recent Actions: ${input.recentActions.length}
${input.recentActions.slice(-10).map(a => `- ${a.action}${a.target ? ` → ${a.target}` : ''}`).join('\n')}

${input.reportHistory?.length ? `
Report History:
${input.reportHistory.map(r => `- ${r.reason}: ${r.outcome}`).join('\n')}
` : 'No report history'}

Assess:
1. Overall risk level
2. Specific concerns
3. Recommended action
4. Monitoring level needed

Focus on patterns, not isolated incidents.`;

        const { output } = await ai.generate({
            prompt,
            output: { schema: BehaviorAnalysisOutputSchema },
        });

        return output ?? {
            riskLevel: 'low' as const,
            concerns: [],
            recommendedAction: 'No action needed',
            monitoringLevel: 'normal' as const,
        };
    }
);

// =============================================================================
// FIREBASE CALLABLE FUNCTIONS
// =============================================================================

export const moderateContent = onCall(
    { maxInstances: 50, timeoutSeconds: 10 },
    async (request) => {
        if (!request.auth) throw new HttpsError('unauthenticated', 'Auth required');
        try {
            return await moderateContentFlow(request.data);
        } catch (error: any) {
            // Fail safe - allow content if moderation fails
            return {
                isSafe: true,
                categories: [],
                suggestedAction: 'allow',
                explanation: 'Moderation unavailable',
            };
        }
    }
);

export const analyzeBehavior = onCall(
    { maxInstances: 10, timeoutSeconds: 30 },
    async (request) => {
        if (!request.auth) throw new HttpsError('unauthenticated', 'Auth required');
        try {
            return await analyzeBehaviorFlow(request.data);
        } catch (error: any) {
            throw new HttpsError('internal', error.message);
        }
    }
);
